concordium-consensus/src/Concordium/Skov/Update.hs

```diff
@@ -461,39 +461,70 @@ doStoreBlock pb@GB.PendingBlock{..} = unlessShutDown $ do
+  maxPendingNum <- rpMaxPendingTransactionNum <$> getRuntimeParameters
+  if pendingNum > maxPendingNum
+  then return ResultUnverifiable
+  else do
+    (_, ur) <- doReceiveTransactionInternal tr slot
+    when (ur == ResultSuccess) $ purgeTransactionTable False =<< currentTime
+    return ur
 
 -- |Add a transaction to the transaction table.  The 'Slot' should be
 -- the slot number of the block that the transaction was received with.
 -- This function should only be called when a transaction is received as part of a block.
 -- The difference from the above function is that this function returns an already existing
 -- transaction in case of a duplicate, ensuring more sharing of transaction data.
-doReceiveTransactionInternal :: (TreeStateMonad pv m) => BlockItem -> Slot -> m (Maybe BlockItem, UpdateResult)
-doReceiveTransactionInternal tr slot =
-        addCommitTransaction tr slot >>= \case
-          Added bi@WithMetadata{..} -> do
-              ptrs <- getPendingTransactions
-              case wmdData of
-                NormalTransaction tx -> do
-                  focus <- getFocusBlock
-                  st <- blockState focus
-                  macct <- getAccount st $! transactionSender tx
-                  nextNonce <- fromMaybe minNonce <$> mapM getAccountNonce macct
-                  -- If a transaction with this nonce has already been run by
-                  -- the focus block, then we do not need to add it to the
-                  -- pending transactions.  Otherwise, we do.
-                  when (nextNonce <= transactionNonce tx) $
-                      putPendingTransactions $! addPendingTransaction nextNonce WithMetadata{wmdData=tx,..} ptrs
-                CredentialDeployment _ ->
-                  putPendingTransactions $! addPendingDeployCredential wmdHash ptrs
-                ChainUpdate cu -> do
-                    focus <- getFocusBlock
-                    st <- blockState focus
-                    nextSN <- getNextUpdateSequenceNumber st (updateType (uiPayload cu))
-                    when (nextSN <= updateSeqNumber (uiHeader cu)) $
-                        putPendingTransactions $! addPendingUpdate nextSN cu ptrs
-              return (Just bi, ResultSuccess)
-          Duplicate tx -> return (Just tx, ResultDuplicate)
-          ObsoleteNonce -> return (Nothing, ResultStale)
+doReceiveTransactionInternal :: (TreeStateMonad pv m, SkovQueryMonad pv m) => BlockItem -> Slot -> m (Maybe BlockItem, UpdateResult)
+doReceiveTransactionInternal tr slot = do
+        -- let maxPendingTransactionsPerAccount = 100
+        -- let maxPendingCredentials = 100
+        maxTimeToExpiry <- durationMillis . rpMaxTimeToExpiry <$> getRuntimeParameters
+        slotTime <- getSlotTimestamp slot
+        let expiry = transactionTimeToTimestamp $ msgExpiry tr
+        if tsMillis (expiry - slotTime) > maxTimeToExpiry
+        then return (Just tr, ResultExpiryTooLate)
```

Comment:
This is problematic since if a transaction is a part of a block we must insert it, and there is no protocol level rule that says that its expiry will be reasonable if it is a part of a block.

So this check should not be here, but only in the `doReceiveTransaction`.